### ISSUE TITLE
Unneeded space after opening parenthesis with AnnotationBear language setting prompt

### ISSUE THREAD

**Issue by underyx** [coala/coala#3164]
```
Please enter a value for the setting "language" ( The programming language of the source code.) needed by AnnotationBear for section "keywords":
```

This is probably a problem inside AnnotationBear but I haven't actually checked where that space comes from.

difficulty newcomer

**Comment by UMANIAX** [comment]
hey can i be assigned this?

**Comment by dracarys09** [comment]
Hi, I would like to be assigned on this.

**Comment by dracarys09** [comment]
@jayvdb I just want to confirm this... I'm supposed to delete the space after opening parenthesis and submit a pull request ?

**Comment by jayvdb** [comment]
Yes.

**Comment by dracarys09** [comment]
@jayvdb   I think I found the cause of this extra space after parenthesis. The reason is the way in which documentation of run() method of AnnotationBear.py
This is the way it is written as of now:  (all bears having this type of doc are showing this extra space)

:param language: 
			The programming language of the source code.

Instead in other bears (eg. SpaceConsistencyBear) the documentation is like:

:param use_spaces:                True if spaces are to be used instead
                                          of tabs.

Bears having this type of documentation are showing output properly (i.e., without extra space)

So should I go ahead and change the doc of AnnotationBear's run() method accordingly... I need your thoughts on this...

**Comment by jayvdb** [comment]
How many bears have this problem?

**Comment by dracarys09** [comment]
All bears which are dependent on AnnotationBear  (QuotesBear, KeywordBear, IndentationBear). However changing only doc of run() method of AnnotationBear.py solves the problem for all these bears.

The same problem is present in CPDBear as well, however this is not dependent of AnnotationBear so to solve this we need to do the same in CPDBear's run() method doc.

Summary : We need to modify one line of documentation of run() method in AnnotationBear.py and                   CPDBear.py
This will resolve this issue.

**Comment by jayvdb** [comment]
could you paste the diff of those two changes here . use three back-ticks before and after, so it is formatted correctly.
changing them is only a temportary workaround. it is probably good enough, but we also need a way to stop it happening again. (maybe that is a separate task.)

**Comment by dracarys09** [comment]
``` this is sparta```

**Comment by dracarys09** [comment]
```diff --git a/bears/general/AnnotationBear.py b/bears/general/AnnotationBear.py
index 5c80026..33b1336 100644
--- a/bears/general/AnnotationBear.py
+++ b/bears/general/AnnotationBear.py
@@ -18,17 +18,18 @@ class AnnotationBear(LocalBear):
         The Bear searches for valid comments and strings and yields their
         ranges as SourceRange objects in HiddenResults.
 
-        :param language:
-            The programming language of the source code.
-        :param coalang_dir:
-            External directory for coalang file.
-        :return:
-            One HiddenResult containing a dictionary with keys being 'strings'
-            or 'comments' and values being a tuple of SourceRanges pointing to
-            the strings and a tuple of SourceRanges pointing to all comments
-            respectively. The ranges do include string quotes or the comment
-            starting separator but not anything before (e.g. when using
-            ``u"string"``, the ``u`` will not be in the source range).
+        :param language:    The programming language of the source code.
+
+        :param coalang_dir: External directory for coalang file.
+
+        :return:            One HiddenResult containing a dictionary with keys
+                            being 'strings' or 'comments' and values being a
+                            tuple of SourceRanges pointing to the strings and
+                            a tuple of SourceRanges pointing to all comments
+                            respectively. The ranges do include string quotes
+                            or the comment starting separator but not anything
+                            before (e.g. when using ``u"string"``, the ``u``
+                            will not be in the source range).
         """
         try:
             lang_dict = LanguageDefinition(language, coalang_dir=coalang_dir)
diff --git a/bears/general/CPDBear.py b/bears/general/CPDBear.py
index 0f79034..48e3851 100644
--- a/bears/general/CPDBear.py
+++ b/bears/general/CPDBear.py
@@ -58,21 +58,19 @@ class CPDBear(GlobalBear):
         For more details see:
         <https://pmd.github.io/pmd-5.4.1/usage/cpd-usage.html>
 
-        :param language:
-            One of the supported languages of this bear.
-        :param minimum_tokens:
-            The minimum token length which should be reported as a duplicate.
-        :param ignore_annotations:
-            Ignore language annotations when comparing text.
-        :param ignore_identifiers:
-            Ignore constant and variable names when comparing text.
-        :param ignore_literals:
-            Ignore number values and string contents when comparing text.
-        :param ignore_usings:
-            Ignore ``using`` directives in C#.
-        :param skip_duplicate_files:
-            Ignore multiple copies of files of the same name and length in
-            comparison.
+        :param language:             One of the supported languages of this
+                                     bear.
+        :param minimum_tokens:       The minimum token length which should be
+                                     reported as a duplicate.
+        :param ignore_annotations:   Ignore language annotations when comparing
+                                     text.
+        :param ignore_identifiers:   Ignore constant and variable names when
+                                     comparing text.
+        :param ignore_literals:      Ignore number values and string contents
+                                     when comparing text.
+        :param ignore_usings:        Ignore ``using`` directives in C#.
+        :param skip_duplicate_files: Ignore multiple copies of files of the
+                                     same name and length in comparison.
         """
         language = language.lower()
 ```

**Comment by dracarys09** [comment]
@jayvdb shall I go ahead and commit the changes and submit a PR ?

**Comment by jayvdb** [comment]
Im offline Atm. Pls ask on gitter

**Comment by Nosferatul** [comment]
@dracarys09 you did one newcomer issue right? If so, move on a low one!

**Comment by dracarys09** [comment]
@Nosferatul yes I did https://github.com/coala/coala-bears/issues/44

I'm currently working on https://github.com/coala/coala-bears/issues/1106 which is a low difficulty issue. But since I already made the changes to this one I thought of submitting a PR. Shall I go ahead and submit?

I was doing this in parallel and I think the changes needed to solve the issue are what I mentioned in my previous comment.

**Comment by Nosferatul** [comment]
If you did the changes, make the pr.

**Comment by dracarys09** [comment]
@Nosferatul As @jayvdb mentioned earlier in his comment https://github.com/coala/coala/issues/3164#issuecomment-265734307 changing them is only a temporary workaround. it is probably good enough, but we also need a way to stop it happening again. (maybe that is a separate task.)
As suggested by @sils the problem might be the parsing logic in DocumentationExtraction.py 
I'm working on that but this temporary needs this issue to be raised in coala-bears to submit PR there

**Comment by Nosferatul** [comment]
That makes sens

**Comment by Nosferatul** [comment]
DocumentationExtraction.py is in coala

**Comment by dracarys09** [comment]
@Nosferatul DocumentationExtraction.py is in coala but temporary workaround as discussed with jayvdb needs modifying AnnotationBear.py which is in coala-bears.
What should I do ?

**Comment by Nosferatul** [comment]
Ok, if you discussed it, do the changes in AnnotationBear.py as you speaked with jayvdb

**Comment by Nosferatul** [comment]
https://github.com/coala/coala-bears/issues/1109. You are assigned!

**Comment by dracarys09** [comment]
Submitted PR for https://github.com/coala/coala-bears/issues/1109

**Comment by jayvdb** [comment]
On gitter it was established that this is a parsing problem in coala, and so should be fixed there.
[December 8, 2016 11:28 PM](https://gitter.im/coala/coala?at=58498a35be9d43bc63359db8)

**Comment by sils** [comment]
This is not a newcomer issue.

**Comment by dracarys09** [comment]
@sils since you removed the newcomer label... can you assign me this again as the other issue I was working on was blocked due to dependence on something else...

**Comment by sils** [comment]
@dracarys09 this issue is not at all easy, you sure you don't wanna do a low issue first?

**Comment by dracarys09** [comment]
@sils @sils  regarding https://github.com/coala/coala/issues/3164 as we discussed previously... I figured out where the problem is... Actually it is not in DocumentationComment class... Instead the problem is in the way we are extracting metadata for a function... The problem is in ```concat_doc_parts()``` method in ```DocstringMetadata.py```. We can correct this by either changing ```return old + ' ' + new``` to ```return old + '' + new``` in ```DocstringMetadata``` class or by changing ```print(colored(STR_GET_VAL_FOR_SETTING.format(setting_name, arr[0], needed,
                                                 section.name),
                  REQUIRED_SETTINGS_COLOR))``` to ```print(colored(STR_GET_VAL_FOR_SETTING.format(setting_name, arr[0].strip(), needed,
                                                 section.name),
                  REQUIRED_SETTINGS_COLOR))``` in ```require_setting``` method of ```ConsoleInteraction.py```. What are your thoughts on this ?
I think we should do it the second way because if we go with the first option then there will be problems is the param description is in more than one line...However second method would work fine in those cases too... What are your thoughts on this ?

**Comment by sils** [comment]
can we maybe just do a `return (old + ' ' + new).strip()`?

**Comment by dracarys09** [comment]
Yes we can do that too. Actually though implementation wise it is different but essentially this is the same thing because when we do ```arr[0].strip()``` , ```arr[0]``` actually has the value of ```old + ' ' + new```. I've tried using ```(old + ' ' + new).strip()``` and it is working too. Do you want me to change my commit to follow this and push again ?

**Comment by sils** [comment]
you need to fix the bug at the source, just fixing one usecase won't fix
all occurrences!

Sincerely,

Lasse Schuirmann

lasse@schuirmann.net
http://coala.io/ - http://viperdev.io/ - http://gitmate.com/

**Comment by dracarys09** [comment]
The source of this bug is exactly what I mentioned in previous comments. The same problem was occurring in CPDBear (I've mentioned this in previous discussion with @jayvdb) as well, however this fixes all the occurrences of this problem.

**Comment by dracarys09** [comment]
@sils I'll try to explain...

1) ```run_coala()``` is called which in turn calls ```gather_configuration()``` which in turn calls ```fill_settings()```

2) In ```fill_settings()``` if we go to the section where it is mentioned ```#Retrieve needed settings.``` you'll find this line (line 73) -> ```needed = bear.get_non_optional_settings()```. After this value of variable ```needed``` is ```OrderedDict([('language', ('The programming language of the source code.', <class 'str'>))])```.

3) Now what is happening when we do ```needed = bear.get_non_optional_settings()``` is this function in turn returns ```get_metadata().non_optional_params```.

4) Now this ```get_metadata()``` method returns ```FunctionMetadata.from_function(...)```. The params to this ```from_function()``` also contains the ```run()``` method of the bear which we want to use (In this case AnnotationBear or CPDBear etc.) Now in line 166 in FunctionMetadata.py we are fetching the documentation of this run() function by doing ```doc = func.__doc__``` 

5) After this, in line 167 we are doing ```doc_comment = DocstringMetadata.from_docstring(doc)```. After this we are using this ```doc_comment``` to create an OrderedDict and return back non_optional_params (which is needed in step 3). So the origin of this error is exactly present in this ```from_docstring(...)``` method (or more specifically in line 59 of this method). And the proposed changes will resolve this error.

**Comment by sils** [comment]
from_docstring should handle the strip then right?

**Comment by dracarys09** [comment]
Inside from_docstring there is a local function named concat_doc_parts... that should handle the strip correctly (from_docstring is using this local function) which can be done by using ```return (old + ' ' + new).strip()``` in concat_doc_parts as you suggested.
